src/cpp/SearchCriteria.h

```diff
@@ -0,0 +1,125 @@
+//
+// Created by Dmytro Klishch on 10/7/22.
+//
+
+#ifndef KNOWYOURKNOWLEDGEMYFORK_SEARCHCRITERIA_H
+#define KNOWYOURKNOWLEDGEMYFORK_SEARCHCRITERIA_H
+
+#include "doctest.h"
+
+#include <map>
+#include <vector>
+#include <string>
+
+template<class... T>
+using searchCondition = std::function<bool(std::variant<T...>)>;
+
+template<class... T>
+class SearchCriteria {
+public:
+    std::vector<searchCondition<T...>> get(const std::string &component) const {
+        if (!conditions_.count(component)) return std::vector<searchCondition<T...>>();
+        return conditions_.at(component);
+    }
+
+    template<class E>
+    void addEqualToCheck(const std::string &component, const E &value) {
+        conditions_[component].push_back(equalTo(value));
+    }
+
+    template<class E>
+    void addGreaterThanCheck(const std::string &component, const E &value) {
+        conditions_[component].push_back(greaterThan(value));
+    }
+
+    template<class E>
+    void addLessThanCheck(const std::string &component, const E &value) {
+        conditions_[component].push_back(lessThan(value));
+    }
+
+    void addContainsCheck(const std::string &component, const std::string &value) {
+        conditions_[component].push_back(contains(value));
+    }
+
+    void addCheck(const std::string &component, searchCondition<T...> condition) {
+        conditions_[component].push_back(condition);
+    }
+
+private:
+    std::map<std::string, std::vector<searchCondition<T...>>> conditions_;
+
+    searchCondition<T...> contains(const std::string &value) {
+        return [value](std::variant<T...> variant) {
+            return value.find(std::get<std::string>(variant)) != -1;
+        };
+    }
+
+    template<class E>
+    searchCondition<T...> equalTo(const E &value) {
```

Comment:
Should such methods become static methods, or maybe standalone functions? They don't use instance fields of SearchCriteria class?